qualtran/bloqs/generalized_qsp.py

```diff
@@ -284,13 +313,136 @@ def _phi(self) -> Sequence[float]:
     def _lambda(self) -> float:
         return self._qsp_phases[2]
 
+    @cached_property
+    def signal_rotations(self) -> NDArray[SU2RotationGate]:
+        return np.array(
+            [
+                SU2RotationGate(theta, phi, self._lambda if i == 0 else 0)
+                for i, (theta, phi) in enumerate(zip(self._theta, self._phi))
+            ]
+        )
+
     def decompose_from_registers(
         self, *, context: cirq.DecompositionContext, signal, **quregs: NDArray[cirq.Qid]
     ) -> cirq.OP_TREE:
-        assert len(signal) == 1
         signal_qubit = signal[0]
 
-        yield SU2RotationGate(self._theta[0], self._phi[0], self._lambda).on(signal_qubit)
-        for theta, phi in zip(self._theta[1:], self._phi[1:]):
-            yield self.U.on_registers(**quregs).controlled_by(signal_qubit, control_values=[0])
-            yield SU2RotationGate(theta, phi, 0).on(signal_qubit)
+        num_inverse_applications = self.negative_power
+
+        yield self.signal_rotations[0].on(signal_qubit)
+        for signal_rotation in self.signal_rotations[1:]:
+            if num_inverse_applications > 0:
+                # apply C-U^\dagger
+                yield self.U.adjoint().on_registers(**quregs).controlled_by(signal_qubit)
+                num_inverse_applications -= 1
+            else:
+                # apply C[0]-U
+                yield self.U.on_registers(**quregs).controlled_by(signal_qubit, control_values=[0])
+            yield signal_rotation.on(signal_qubit)
+
+        for _ in range(num_inverse_applications):
+            yield self.U.adjoint().on_registers(**quregs)
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        degree = len(self.P)
+        return {
+            (self.U.adjoint().controlled(), min(degree, self.negative_power)),
+            (self.U.controlled(ctrl_spec=CtrlSpec(cvs=0)), max(0, degree - self.negative_power)),
+            (self.U.adjoint(), max(0, self.negative_power - degree)),
+        } | {(rotation, 1) for rotation in self.signal_rotations}
+
+
+@frozen
+class HamiltonianSimulationByGQSP(GateWithRegisters):
```

Comment:
Let's move out the "support negative powers" part in a separate PR and merge it, since it's working and the tests pass.